magenta/models/polyphonic_rnn/polyphonic_rnn_lib.py

```diff
@@ -0,0 +1,1377 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utilities for polyphonic RNN."""
+
+from collections import OrderedDict
+import copy
+import os
+import re
+import uuid
+
+# internal imports
+
+import numpy as np
+from scipy import linalg
+import tensorflow as tf
+
+import magenta.music as mm
+from magenta.protobuf import music_pb2
+
+
+##
+# begin metautils
+##
+
+
+def shape(x):
+  # Get shape of Variable through hacky hacky string parsing
+  shape_tup = repr(tf.Print(x, [tf.shape(x)])).split('shape=')[1]
```

Comment:
There must be a better way to get numeric shapes out of TF, but I couldn't find it. We talked about a way, but when I tried it I couldn't get the information I needed. Maybe there is a better way from the public API that I missed?